File: app/src/androidTest/java/com/przemolab/oknotifier/matchers/Matchers.java
package com.przemolab.oknotifier.matchers;

import android.content.res.Resources;
import android.graphics.drawable.ColorDrawable;
import android.support.test.espresso.intent.Checks;
import android.support.test.espresso.matcher.BoundedMatcher;
import android.support.v7.widget.RecyclerView;
import android.view.View;

import com.przemolab.oknotifier.R;

import org.hamcrest.Description;
import org.hamcrest.Matcher;
import org.hamcrest.TypeSafeMatcher;

public class Matchers {

    public static Matcher<View> isSubscribed(Resources resources) {
        return hasBackgroundColor(resources.getColor(R.color.lightGreen));
    }

    public static Matcher<View> isNotSubscribed(Resources resources) {
        return hasBackgroundColor(resources.getColor(R.color.lightGrey));
    }

    public static RecyclerViewMatcher withRecyclerView(final int recyclerViewId) {
        return new RecyclerViewMatcher(recyclerViewId);
    }

    private static Matcher<View> hasBackgroundColor(final int color) {

        Checks.checkNotNull(color);

        return new BoundedMatcher<View, View>(View.class) {

            @Override
            public boolean matchesSafely(View view) {
                return color == ((ColorDrawable) view.getBackground()).getColor();
            }

            @Override
            public void describeTo(Description description) {
                description.appendText("has background with color: " + color);
            }
        };
    }
}

File: app/src/androidTest/java/com/przemolab/oknotifier/matchers/Matchers.kt
package com.przemolab.oknotifier.matchers

import android.content.Context
import android.graphics.drawable.ColorDrawable
import android.support.test.espresso.intent.Checks
import android.support.test.espresso.matcher.BoundedMatcher
import android.support.v4.content.ContextCompat
import android.view.View

import com.przemolab.oknotifier.R

import org.hamcrest.Description
import org.hamcrest.Matcher

object Matchers {

    @JvmStatic
    fun isSubscribed(context: Context): Matcher<View> {
        return hasBackgroundColor(ContextCompat.getColor(context, R.color.lightGreen))
    }

    @JvmStatic
    fun isNotSubscribed(context: Context): Matcher<View> {
        return hasBackgroundColor(ContextCompat.getColor(context, R.color.lightGrey))
    }

    @JvmStatic
    fun withRecyclerView(recyclerViewId: Int): RecyclerViewMatcher {
        return RecyclerViewMatcher(recyclerViewId)
    }

    private fun hasBackgroundColor(color: Int): Matcher<View> {

        Checks.checkNotNull(color)

        return object : BoundedMatcher<View, View>(View::class.java) {

            public override fun matchesSafely(view: View): Boolean {
                return color == (view.background as ColorDrawable).color
            }

            override fun describeTo(description: Description) {
                description.appendText("has background with color: $color")
            }
        }
    }
}
